File: src/api/api_security.ts
export interface ApiEndpointData {
  id: number;
  project_id: number;
  path: string;
  method: string;
  summary: string | null;
  operation_id: string | null;
  auth_status: string;
  auth_type: string | null;
  rate_limit_status: string;
  request_validation_status: string;
  sensitive_data_fields: string | null;
  bola_status?: string | null;
  mass_assignment_status?: string | null;
  risk_score: number;
  risk_level: string;
  discovered_via: string;
  created_at: string;
  updated_at: string;
}

export interface ApiSecuritySummaryData {
  total_endpoints: number;
  critical_endpoints: number;
  high_endpoints: number;
  medium_endpoints: number;
  low_endpoints: number;
  info_endpoints: number;
  unauthenticated_endpoints: number;
  sensitive_data_endpoints: number;
  unconstrained_validation_endpoints: number;
  bola_risk_endpoints?: number;
  mass_assignment_endpoints?: number;
  missing_rate_limit_endpoints?: number;
  total_api_findings: number;
  open_api_findings: number;
}

export interface OpenApiIngestResponseData {
  project_id: number;
  source_type: string;
  status: string;
  endpoints_count: number;
  message: string;
}

const API_BASE_URL = (import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000').replace(/\/$/, '');

const getErrorMessage = (response: Response) => {
  if (response.status === 0) {
    return 'The Kyptic backend is unavailable. Start it on http://localhost:8000 and try again.';
  }
  return `The Kyptic backend returned an error (${response.status}). Please try again.`;
};

export const ingestOpenApi = async (projectId: number, file: File): Promise<OpenApiIngestResponseData> => {
  const formData = new FormData();
  formData.append('file', file);

  const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}/ingest/openapi`, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    const errorText = await response.text();
    try {
      const errorJson = JSON.parse(errorText);
      throw new Error(errorJson.detail || getErrorMessage(response));
    } catch {
      throw new Error(errorText || getErrorMessage(response));
    }
  }

  return response.json() as Promise<OpenApiIngestResponseData>;
};

export const analyzeApiSecurity = async (projectId: number): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}/api-security/analyze`, {
    method: 'POST',
  });

  if (!response.ok) {
    const errorText = await response.text();
    try {
      const errorJson = JSON.parse(errorText);
      throw new Error(errorJson.detail || getErrorMessage(response));
    } catch {
      throw new Error(errorText || getErrorMessage(response));
    }
  }

  return response.json();
};

export const fetchApiEndpoints = async (projectId: number): Promise<ApiEndpointData[]> => {
  const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}/api-security/endpoints`);
  if (!response.ok) {
    throw new Error(getErrorMessage(response));
  }
  return response.json() as Promise<ApiEndpointData[]>;
};

export const fetchApiEndpointDetail = async (projectId: number, endpointId: number): Promise<ApiEndpointData> => {
  const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}/api-security/endpoints/${endpointId}`);
  if (!response.ok) {
    throw new Error(getErrorMessage(response));
  }
  return response.json() as Promise<ApiEndpointData>;
};

export const fetchApiSecuritySummary = async (projectId: number): Promise<ApiSecuritySummaryData> => {
  const response = await fetch(`${API_BASE_URL}/api/projects/${projectId}/api-security/summary`);
  if (!response.ok) {
    throw new Error(getErrorMessage(response));
  }
  return response.json() as Promise<ApiSecuritySummaryData>;
};

File: backend/app/schemas/api_security.py
from datetime import datetime
from pydantic import BaseModel, ConfigDict


class ApiEndpointResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    path: str
    method: str
    summary: str | None = None
    operation_id: str | None = None
    auth_status: str
    auth_type: str | None = None
    rate_limit_status: str
    request_validation_status: str
    sensitive_data_fields: str | None = None
    bola_status: str = "NONE"
    mass_assignment_status: str = "NONE"
    dast_status: str = "UNTESTED"
    risk_score: int
    risk_level: str
    discovered_via: str
    created_at: datetime
    updated_at: datetime


class ApiSecuritySummaryResponse(BaseModel):
    total_endpoints: int
    critical_endpoints: int
    high_endpoints: int
    medium_endpoints: int
    low_endpoints: int
    info_endpoints: int
    unauthenticated_endpoints: int
    sensitive_data_endpoints: int
    unconstrained_validation_endpoints: int
    bola_risk_endpoints: int = 0
    mass_assignment_endpoints: int = 0
    missing_rate_limit_endpoints: int = 0
    total_api_findings: int
    open_api_findings: int

    # Phase 3 DAST Verification Summary Metrics
    verified_vulnerable_endpoints: int = 0
    verified_secure_endpoints: int = 0
    inconclusive_endpoints: int = 0
    untested_endpoints: int = 0
    static_findings_count: int = 0
    dast_findings_count: int = 0
    last_dast_scan_status: str | None = None
    last_dast_scan_at: datetime | None = None


class OpenApiIngestResponse(BaseModel):
    project_id: int
    source_type: str
    status: str
    endpoints_count: int
    message: str


class DastTargetConfigRequest(BaseModel):
    api_target_url: str
    api_dast_enabled: bool = True
    api_auth_type: str = "NONE"  # NONE, BEARER, API_KEY, BASIC
    api_auth_header_name: str | None = "Authorization"
    api_auth_token: str | None = None


class DastTargetConfigResponse(BaseModel):
    project_id: int
    api_target_url: str | None = None
    api_dast_enabled: bool = False
    api_auth_type: str | None = "NONE"
    api_auth_header_name: str | None = "Authorization"
    is_ssrf_safe: bool = False
    message: str


class DastTestConnectionResponse(BaseModel):
    status: str  # SUCCESS, FAILED, BLOCKED
    target_url: str
    status_code: int | None = None
    elapsed_ms: float = 0.0
    message: str
